lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -461,13 +461,32 @@ static SILValue reapplyFunctionConversion(
       context.addDifferentiableFunctionInstToWorklist(dfi);
       newArgs.back() = dfi;
     }
-    // If new function's generic signature is specified, use it to create
-    // substitution map for reapplied `partial_apply` instruction.
-    auto substMap = !newFuncGenSig
-        ? pai->getSubstitutionMap()
-        : SubstitutionMap::get(
-              newFuncGenSig, QuerySubstitutionMap{pai->getSubstitutionMap()},
-              LookUpConformanceInModule(builder.getModule().getSwiftModule()));
+    // Compute substitution map for reapplying `partial_apply`.
+    // - If reapplied functoin is not polymorphic, use empty substitution map
+    //   regardless of the original `partial_apply`'s substitution map.
+    //   - This case is triggered for reapplying `partial_apply` where `newFunc`
+    //     is a `differentiability_witness_function` where the witness generic
+    //     signature has all concrete parameters while the original function's
+    //     generic signature does not. In this case, the original function type
+    //     is polymorphic while derivative function types are not (specialized
+    //     with concrete types from same-type requirements).
+    // - Otherwise, if `newFuncGenSig` is not specified, use the original
+    //   `partial_apply`'s substitution map.
+    // - Otherwise, if `newFuncGenSig` is specified, combine it with the
+    //   original `partial_apply`'s substitution map.
+    SubstitutionMap substMap;
+    if (innerNewFunc->getType().castTo<SILFunctionType>()->isPolymorphic()) {
+      if (!newFuncGenSig) {
+        substMap = pai->getSubstitutionMap();
+      } else {
+        substMap = SubstitutionMap::get(
+              newFuncGenSig,
```

Comment:
Fix indentation.

Reply:
Thanks for the catch! Done in https://github.com/apple/swift/pull/28772/commits/e2986239c13472e37dd765744e7075a96b7c92e3.